prog/dftb+/lib_dftb/dispdftd4.F90

```diff
@@ -0,0 +1,1290 @@
+          dEdq(iAt1) = dEdq(iAt1) - dc9dq1*dEr
+          dEdq(iAt2f) = dEdq(iAt2f) - dc9dq2*dEr
+          dEdq(iAt3f) = dEdq(iAt3f) - dc9dq3*dEr
+        end do
+      end do
+    end do
+
+  end subroutine threeBodyDispersionGradient
+
+
+  !> Driver for the calculation of DFT-D4 dispersion related properties.
+  subroutine dispersionEnergy(calculator, nAtom, coords, species, neigh, img2CentCell, recPoint,&
+      & energies, gradients, stress, volume, parEwald)
+
+    !> DFT-D dispersion model.
+    type(DftD4Calculator), intent(in) :: calculator
+
+    !> Nr. of atoms (without periodic images)
+    integer, intent(in) :: nAtom
+
+    !> Coordinates of the atoms (including images)
+    real(dp), intent(in) :: coords(:, :)
+
+    !> Species of every atom.
+    integer, intent(in) :: species(:)
+
+    !> Updated neighbour list.
+    type(TNeighbourList), intent(in) :: neigh
+
+    !> Mapping into the central cell.
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Contains the points included in the reciprocal sum.
+    !  The set should not include the origin or inversion related points.
+    real(dp), allocatable, intent(in) :: recPoint(:, :)
+
+    !> Parameter for Ewald summation.
+    real(dp), intent(in), optional :: parEwald
+
+    !> Updated energy vector at return
+    real(dp), intent(out) :: energies(:)
+
+    !> Updated gradient vector at return
+    real(dp), intent(out) :: gradients(:, :)
+
+    !> Upgraded stress
+    real(dp), intent(out), optional :: stress(:, :)
+
+    !> Volume, if system is periodic
+    real(dp), intent(in), optional :: volume
+
+    integer, allocatable :: nNeigh(:)
+
+    real(dp) :: sigma(3, 3)
+    real(dp) :: vol, parEwald0
+    real(dp), allocatable :: cn(:), dcndr(:, :, :), dcndL(:, :, :)
+    real(dp), allocatable :: q(:), dqdr(:, :, :), dqdL(:, :, :)
+
+    if (present(volume)) then
+      vol = volume
+    else
+      vol = 0.0_dp
+    end if
+
+    if (present(parEwald)) then
+       parEwald0 = parEwald
+    else
+       parEwald0 = 0.0_dp
+    end if
+
+    energies(:) = 0.0_dp
+
+    allocate(cn(nAtom), dcndr(3, nAtom, nAtom), dcndL(3, 3, nAtom), source=0.0_dp)
+    allocate(nNeigh(nAtom), source=0)
+
+    call getNrOfNeighboursForAll(nNeigh, neigh, calculator%cutoffCount)
+
+    call getCoordinationNumber(nAtom, coords, species, nNeigh, neigh%iNeighbour, &
+        & neigh%neighDist2, img2CentCell, calculator%covalentRadius, &
+        & calculator%electronegativity, .false., cn, dcndr, dcndL)
+    call cutCoordinationNumber(nAtom, cn, dcndr, dcndL, cn_max=8.0_dp)
+
+    allocate(q(nAtom), dqdr(3, nAtom, nAtom), dqdL(3, 3, nAtom))
+
+    call getNrOfNeighboursForAll(nNeigh, neigh, calculator%cutoffEwald)
+
+    call getEEQCharges(nAtom, coords, species, nNeigh, neigh%iNeighbour, neigh%neighDist2,&
+        & img2CentCell, recPoint, parEwald0, vol, calculator%chi, calculator%kcn, calculator%gam,&
+        & calculator%rad, cn, dcndr, dcndL, qAtom=q, dqdr=dqdr, dqdL=dqdL)
```

Comment:
What is the point of passing `dcndr` and `dcndL` here? The corresponding variables (`stress` and `gradients`) are in `getEEQCharges()` `intent(out)`, so their values should not be used. And the next `getCoordinationNumber()` call rewrites them again, without using them.

Reply:
They are needed to generate `dqdr` and `dqdL`. See here:

https://github.com/dftbplus/dftbplus/blob/5a3a45ce7929cb552bb24bbe3aefc51dbddf8066/prog/dftb%2B/lib_dftb/encharges.F90#L647-L650

Reply:
This is always done, even if `gradients` and `stress` are not present.

Reply:
Sorry, I mixed up the correspondence between dummy arguments and actual ones. You are right.